frontend/src/AppState/PostFeedState.ts

```diff
@@ -0,0 +1,143 @@
+import {getCachedValue, setCachedValue} from "../API/use/useCache";
+import {PostInfo} from "../Types/PostInfo";
+import {action, computed, makeObservable, observable, runInAction} from "mobx";
+import APIHelper from "../API/APIHelper";
+
+export enum PostFeedStateType {
+    Loading,
+    Ready,
+    Error
+}
+
+type GetPostsResult = {
+    posts: PostInfo[];
+    total: number;
+}
+
+export abstract class PostFeedStateBase {
+
+    protected readonly cacheKeyPart: any[];
+    protected readonly api: APIHelper;
+    private readonly cacheName;
+
+    @observable
+    page: number = 1;
+
+    @observable
+    perPage: number;
+
+    @observable
+    pages: number = 0;
+
+    @observable
+    status: PostFeedStateType = PostFeedStateType.Loading;
+
+    readonly posts = observable.array<PostInfo>([]);
+
+    constructor(api: APIHelper, cacheName:string, cacheKeyPart: any[], perPage: number) {
+        this.cacheName = cacheName;
+        this.api = api;
+        this.cacheKeyPart = cacheKeyPart;
+        this.perPage = perPage;
+        this.loadFromCache();
+        makeObservable(this);
+    }
+
+    abstract getPosts(): Promise<GetPostsResult>;
+
+    @computed
+    get loading(): boolean {
+        return this.status === PostFeedStateType.Loading;
+    }
+
+    @computed
+    get error(): boolean {
+        return this.status === PostFeedStateType.Error;
+    }
+
+    @action
+    protected setPage(page: number) {
+        this.page = page;
+    }
+
+    @action
+    protected setPages(pages: number) {
+        this.pages = pages;
+    }
+
+    @action
+    protected setStatus(status: PostFeedStateType) {
+        this.status = status;
+    }
+
+    @action
+    protected setPosts(posts: PostInfo[]) {
+        this.posts.replace(posts);
+    }
+
+    @action // won't be needed after all PostFeeds are transitioned to MobX
+    updatePost = (id: number, partial: Partial<PostInfo>) => {
+        const index = this.posts.findIndex(post => post.id === id);
+        if (index !== -1) {
+            Object.assign(this.posts[index], partial);
+        }
+    };
+
+    private loadFromCache() {
+        const cachedPosts: PostInfo[] | undefined = getCachedValue(this.cacheName, this.cacheKey) ||
+            getCachedValue('feed-fallback', []); // display something (it's blurred anyway)
```

Comment:
> The whole idea of this cache itself was to fix issue with navigation 

Then it doesn’t work as intended (or at least it didn’t when I last saw it. As far as I understand the current logic, the cache stores a single page for “feed” — the last page the user saw. Which means only current page is present in the cache (not the previous). Due to another bug (I assume it’s a bug), the previous page is fetched from cache and shown during loading, along with «Загрузка…». E.g. currently, if you’re switching back from page 2 to page 1, the page 2 content will be shown while page 1 is loading.

In this demo I used the cache to display “somewhat relevant” blurred post feed as a loading indicator. I remember @4vanger suggested  to use blurring instead of the loading indicator. 

But I’m not insisting on this approach, I can implement the previous page cache approach that you’ve outlined (however I’m not sure that it’s a good idea to cache previous page instead of requesting the actual information).

Reply:
You're right. Maybe I didn't explain wall. It was not supposed to cache all requests and was used only for one most useful case:
- open post from feed
- navigate back
- open next post from the same feed

Reply:
Let’s discuss this in the chat. I have doubts that it’s a good idea to show interactive posts to the user while page could be changed at any moment (when loading finishes).

Reply:
One potential approach to discuss could be the following: to load prev/next page in the background and, if the page is present in the cache, show it immediately (without even trying to load the current version from the server).

Reply:
I went ahead and simplified the whole thing.

Current logic: 
* only the latest (current) page is stored in the cache (same cache is shared between all feeds)
* on page loading, whatever is in the cache is shown (blurred)
  * if cache is empty, "loading..." text is shown
* when going back, same logic (as above) applies. Rationale: 
  * going back to the previous page within the same feed is a rare use case, no reason to support it (i.e., page navigation usually happens only forward)
  * when switching feeds it's reasonable to expect the freshly loaded data